source/base/ThreadPool.ooc

```diff
@@ -1,111 +1,162 @@
 use ooc-collections
-
+use ooc-base
 import structs/LinkedList
 import threading/Thread
+import os/Time
 
-ThreadJob: class {
-	_body: Func
-	init: func (=_body)
-	_execute: func { this _body() }
-	run: virtual func {
-		this _execute()
-		this free()
-	}
-}
-
-SynchronizedThreadJob: class extends ThreadJob {
-	_finishedCondition := WaitCondition new()
-	_mutex := Mutex new()
-	_finished := false
-	finished ::= this _finished
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
-	free: override func {
-		this _mutex lock()
-		if (!this _finished) {
-			this _freeOnCompletion = true
-			this _mutex unlock()
-		} else {
+_Task: abstract class {
+	_mutexUpdateTime: static Int = 1
+	_state := _PromiseState Unfinished
+	_freeDirectly: Bool
+	_mutex: Mutex
+	mutex ::= this _mutex
+	run: abstract func (mutex: Mutex)
+	wait: func -> Bool {
+		status := false
+		while (!status) {
+			this _mutex lock()
+			if (this _state != _PromiseState Unfinished) {
+				status = (this _state == _PromiseState Finished)
+				this _mutex unlock()
+				break
+			}
 			this _mutex unlock()
-			this _mutex destroy()
-			this _finishedCondition free()
-			super()
+			Time sleepMilli(This _mutexUpdateTime)
```

Comment:
I will let him speak for himself, but we talked about this earlier and he seemed to agree with this approach for the things we're using it for (this was also written before WaitCondition was finished and platform independent - I believe it is now?).